_posts/2022-04-26-type-based-optimizations-in-the-jit.md

```diff
@@ -0,0 +1,804 @@
+    X + Y.
+```
+
+The BEAM code looks like:
+
+```
+    {gc_bif,'*',{f,0},2,[{x,0},{integer,2}],{x,0}}.
+    {gc_bif,'*',{f,0},2,[{x,1},{integer,2}],{x,1}}.
+    {gc_bif,'+',{f,0},2,[{tr,{x,0},number},{tr,{x,1},number}],{x,0}}.
+    return.
+```
+
+Types are propagated from arithmetic instructions to other arithmetic
+instructions. Because the result of `*` (if it succeeds) is a number
+(integer or float), the operands for the `+` instruction now have the
+type `number`.
+
+Based on our experience of adding types to the `<` operator, we might
+guess that we would save only one instruction in the type test. We
+would be right:
+
+```
+# simplified test for small operands since both are numbers
+    mov eax, esi
+    and eax, edx
+    test al, 1
+    short je L22
+```
+
+Returning to the simpler example with addition and no multiplication,
+let's add a guard to ensure that `X` and `Y` are integers:
+
+```
+add3(X, Y) when is_integer(X), is_integer(Y) ->
+    X + Y.
+```
+
+That results in the following BEAM code:
+
+
+```
+    {test,is_integer,{f,5},[{x,0}]}.
+    {test,is_integer,{f,5},[{x,1}]}.
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,any}},{tr,{x,1},{t_integer,any}}],
+            {x,0}}.
+    return.
+```
+
+The types for both operands are now `{t_integer,any}`. However, that
+will still result in the same simplified four-instruction sequence for
+testing small integers, because the integers might not fit in 60 bits.
+
+Clearly, based on our experience with `is_lt`, we will need to establish
+a range for `X` and `Y`. A reasonable way to do that would be:
+
+```
+add4(X, Y) when is_integer(X), 0 =< X, X < 16#400,
+                is_integer(Y), 0 =< Y, Y < 16#400 ->
+    X + Y.
+```
+
+Unfortunately, because of limitations in the compiler's value range analysis,
+the types for the `+` operator will **not** improve:
+
+```
+    {test,is_integer,{f,19},[{x,0}]}.
+    {test,is_ge,{f,19},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
+    {test,is_lt,{f,19},[{tr,{x,0},{t_integer,any}},{integer,1024}]}.
+    {test,is_integer,{f,19},[{x,1}]}.
+    {test,is_ge,{f,19},[{tr,{x,1},{t_integer,any}},{integer,0}]}.
+    {test,is_lt,{f,19},[{tr,{x,1},{t_integer,any}},{integer,1024}]}.
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,any}},{tr,{x,1},{t_integer,any}}],
+            {x,0}}.
+    return.
+```
+
+To add insult to injury, the first 6 instructions cannot be simplified
+by the JIT because there is not sufficient type information. That is,
+the `is_lt` and `is_ge` instruction will comprise 12 instructions each.
+
+We hope to improve the type analysis and optimizations in OTP 26
+and generate better code for this example.
+
+Meanwhile, while we wait for OTP 26, there is a way in
+OTP 25 to write an equivalent guard that will result in
+much more efficient code **and** establish known ranges for `X` and
+`Y`:
+
+```
+add5(X, Y) when X =:= X band 16#3FF,
```

Comment:
I hope we are not suggesting the user to optimize with the use of guards knowing how the JIT currently can optimize.
I suggest that we introduce a guard BIF `is_integer_range/2` or similar in OTP 26 to allow for better optimization and for nicer looking code. It would then be obvious that it is a comparision between integers and nothing else.

Reply:
> I hope we are not suggesting the user to optimize with the use of guards knowing how the JIT currently can optimize.

I suppose I could add some discouraging words that it should be avoided.

>  I suggest that we introduce a guard BIF `is_integer_range/2` or similar in OTP 26 to allow for better optimization and for nicer looking code. It would then be obvious that it is a comparision between integers and nothing else.

Totally agree. We should do that.

Reply:
I have added a slight discouragement and also mentioned that we are considering adding a new guard BIF for range checking.